File: day0008/SortIntByBit.java
*****************Auther >> Akash (The Robust Coder)*******************
/*

Approch  :
Use Recursively Bubble Sort for sorting and Count bits Using Integer.bitCount Inbuilt function  :)




*/


class Solution {
    public int[] sortByBits(int[] arr) {
        return bubbleSort(arr, arr.length);
        
    }
    public int[] bubbleSort(int[] arr, int len){
        int i , temp;
        for(i = 0; i < len-1; i++){
            int bc1 = Integer.bitCount(arr[i]);
            int bc2 = Integer.bitCount(arr[i+1]);
            
            //Comparing Elements
            if((bc1 > bc2) || ((bc1 == bc2) && (arr[i] > arr[i+1]))) {
                temp = arr[i];
                arr[i] = arr[i+1];
                arr[i+1] = temp; 
            }
        }
        if(len == 0) return arr;
        //sort again whenevr we got arr lent 0
        return bubbleSort(arr, len-1);
    }
}

File: day0008/Sort_intbybit.cpp
/*
  ******************auther- Anand Keshari********************
  problem link - https://leetcode.com/problems/sort-integers-by-the-number-of-1-bits/
*/
int bit_count(int a)
    {
        unsigned int count=0;
        while(a)
        {
            if(a & 1)
                count++;
            a=a>>1;
        }
        return count;
    }
bool cmp(int a1, int a2 )
    {
        if(bit_count(a1)<bit_count(a2))
            return true;
        else if(bit_count(a1)>bit_count(a2))
            return false;
        return a1<a2;
    }
class Solution {
public:
    vector<int> sortByBits(vector<int>& arr) {
       sort(arr.begin(),arr.end(),cmp);
        return arr;
    }
    
    
};
